microcosm_pubsub/context.py

```diff
@@ -5,17 +5,45 @@
 
 from microcosm.api import defaults, typed
 from microcosm.config.types import boolean
+from microcosm_logging.decorators import logger
 
+from microcosm_pubsub.errors import TTLExpired
 
-def sqs_message_context(message_dct, **kwargs):
-    context = message_dct.get("opaque_data", dict())
 
-    # If there is a uri add it to the context
-    if message_dct.get("uri"):
-        context["uri"] = message_dct.get("uri")
+@logger
+class SQSMessageContext:
+    """
+    The message context controls what data you want to associate
+    with your daemon handler context, e.g. some combination of keys from the message or
+    additional metadata fetched from elsewhere.
 
-    context.update(**kwargs)
-    return context
+    """
+    def __init__(self, graph):
+        self.enable_ttl = graph.config.sqs_message_context.enable_ttl
+        self.initial_ttl = graph.config.sqs_message_context.initial_ttl
+
+    def __call__(self, message_dct, **kwargs):
+        context = message_dct.get("opaque_data", dict())
+
+        # If there is a uri add it to the context
+        if message_dct.get("uri"):
+            context["uri"] = message_dct.get("uri")
+
+        if self.enable_ttl:
+            try:
+                ttl = int(context["X-Request-Ttl"])
+            except KeyError:
+                ttl = self.initial_ttl
+            if ttl == 0:
+                self.logger.warning(
+                    f"Error handling SQS message - TTL expired",
+                    extra=context,
+                )
+                raise TTLExpired(extra=context)
+            context["X-Request-Ttl"] = str(ttl - 1)
```

Comment:
What's this for?

Reply:
the `str`?  because the retrieved message contains only strings (and not numbers)

Reply:
I mean the `- 1`.